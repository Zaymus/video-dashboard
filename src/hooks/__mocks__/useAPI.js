// Default frozen result for normal tests
export const defaultResult = Object.freeze({
  result: {
    items: [
      { contentDetails: { id: '1' }, snippet: { title: 'Video 1' } },
      { contentDetails: { id: '2' }, snippet: { title: 'Video 2' } },
      { contentDetails: { id: '3' }, snippet: { title: 'Video 3' } }
    ],
    nextPageToken: 'ABC123'
  },
  isLoading: false,
  hasError: null
});

// Helper mocks for different test scenarios
export const loadingMock = Object.freeze({
  result: null,
  isLoading: true,
  hasError: null
});

export const errorMock = Object.freeze({
  result: null,
  isLoading: false,
  hasError: 'API error'
});

export const moreVideosMock = Object.freeze({
  result: {
    items: [
      { contentDetails: { id: '4' }, snippet: { title: 'Video 4' } },
      { contentDetails: { id: '5' }, snippet: { title: 'Video 5' } },
      { contentDetails: { id: '6' }, snippet: { title: 'Video 6' } }
    ],
    nextPageToken: 'DEF456'
  },
  isLoading: false,
  hasError: null
});

// Default export returns a controlled mock (can override in tests)
let currentMock = defaultResult;

export const setUseAPIMock = (mock) => {
  currentMock = mock || defaultResult;
};

export default () => currentMock;