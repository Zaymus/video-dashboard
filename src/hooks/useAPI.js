import axios from 'axios';
import { useState, useEffect } from 'react';
import { isURLValid } from '../utils/constants';

const useAPI = (endpoint, config) => {
  const [isLoading, setIsLoading] = useState(false);
  const [hasError, setHasError] = useState(false);
  const [result, setResult] = useState(null);

  useEffect(() => {
    if (!endpoint || !isURLValid(endpoint)) {
      setHasError(`Invalid Endpoint: ${endpoint}`);
      return;
    };

    let isCancelled = false;
    setIsLoading(true);
    setHasError(false);

    axios(endpoint, config)
      .then(res => {
        if(res.status !== 200) {
          throw new Error(`HTTP Error: Status: ${res.status}`);
        }
        return res.data;
      })
      .then(data => {
        if (!isCancelled) {
          setResult(data);
          setIsLoading(false);
        }
      })
      .catch(err => {
        if (!isCancelled) {
          setHasError(err.message || 'Something went wrong when calling API');
          setIsLoading(false);
        }
      });

      return () => {
        isCancelled = true;
      }
  }, [endpoint, config]);

  return { result, isLoading, hasError };
} 

export default useAPI;