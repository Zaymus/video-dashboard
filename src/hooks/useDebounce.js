import { useEffect, useState, useRef } from 'react';

const useDebounce = (input, delay) => {
  const [debouncedValue, setDebouncedValue] = useState(undefined);
  const callbackRef = useRef();
  const timeoutRef = useRef();

  useEffect(() => {
    if (typeof input === 'function') {
      callbackRef.current = input;
    }
  }, [input]);

  const debouncedFn = useRef();
  if (typeof input === 'function' && !debouncedFn.current) {
    debouncedFn.current = () => {
      clearTimeout(timeoutRef.current);
      timeoutRef.current = setTimeout(() => {
        callbackRef.current();
      }, delay);
    };
  }

  useEffect(() => {
    if (typeof input !== 'function') {
      const timerID = setTimeout(() => setDebouncedValue(input), delay);
      return () => clearTimeout(timerID);
    }
  }, [input, delay]);

  return typeof input === 'function' ? debouncedFn.current : debouncedValue;
}

export default useDebounce;