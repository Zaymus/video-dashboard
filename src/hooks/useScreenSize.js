import { useState, useEffect } from 'react';
import { SCREEN_SIZES } from '../utils/constants';

const useScreenSize = () => {
  const [ screenSize, setScreenSize ] = useState(null);

  const resizeHandler = () => {
    const width = window.innerWidth;

    if (width <= 375) {
      setScreenSize(SCREEN_SIZES.MOBILE_SMALL);
    } else if (width <= 425) {
      setScreenSize(SCREEN_SIZES.MOBILE_LARGE);
    } else if (width <= 768) {
      setScreenSize(SCREEN_SIZES.TABLET);
    } else {
      setScreenSize(SCREEN_SIZES.DESKTOP);
    }
  }

  useEffect(() => {
    if (!window) return;
      
    resizeHandler();
    window.addEventListener('resize', resizeHandler);

    return () => {
      window.removeEventListener('resize', resizeHandler);
    }
  }, []);

  return { screenSize, SCREEN_SIZES };
};

export default useScreenSize;