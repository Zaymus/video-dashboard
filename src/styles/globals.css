:root {
  --primary: #1E90FF;
  --secondary: #FF6B6B;
  --background-dark: #121212;
  --surface-dark: #1F1F1F;
  --text-primary-dark: #FFFFFF;
  --text-secondary-dark: #B3B3B3;
  --border-dark: #2C2C2C;
  --success: #2ECC71;
  --error: #E74C3C;
  --warning: #F1C40F;
  --background-light: #F9F9F9;
  --surface-light: #FFFFFF;
  --text-primary-light: #1A1A1A;
  --text-secondary-light: #4F4F4F;
  --border-light: #E0E0E0;

  --videos-gutter-margin: 30px;
  --videos-per-row: 5;
  --videos-usable-width: calc(100% - var(--videos-gutter-margin) * 2);
}

body {
  background-color: var(--background-dark);
  margin: 0;
  color: var(--text-primary-dark);
  font-size: 16pt;
}

#root {
  min-height: 100vh;
}