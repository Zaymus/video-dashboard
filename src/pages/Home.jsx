import { useEffect, useState } from 'react';
import styled from 'styled-components';
import Loader from '../components/common/Loader';
import useAPI from '../hooks/useAPI';
import { API_ENDPOINTS, POPULAR_VIDEOS_REQUEST } from '../utils/constants';
import useScrollBar from '../hooks/useScrollBottom';

const Container = styled.div`
  margin-top: 75px;
  padding: 10px 20px;
`;

const Home = () => {
  const [videos, setVideos] = useState([]);
  const [nextPageToken, setNextPageToken] = useState(null);
  const [requestConfig, setRequestConfig] = useState(POPULAR_VIDEOS_REQUEST);
  const { result, isLoading, hasError } = useAPI(API_ENDPOINTS.getVideos, requestConfig);
  const isAtBottom = useScrollBar(20);

  useEffect(() => {
    if (result) {
      setNextPageToken(result.nextPageToken);
      setVideos(prevVideos => {
        return prevVideos?.length > 0 ? [...prevVideos, ...result.items] : [...result.items];
      });
    }
  }, [result])

  useEffect(() => {
    if (isAtBottom) {
      setRequestConfig(prevConfig => ({
        ...prevConfig,
        params: {
          ...prevConfig.params,
          pageToken: nextPageToken
        }
      }))
    }
  }, [isAtBottom, nextPageToken]);

  return (
    <Container id="homeContainer">
        {
          videos.length > 0 && videos.map(video => {
            return <p key={video.contentDetails.id}>{video.snippet.title}</p>;
          })
        }
        { isLoading == true && <Loader /> }
        { hasError && <p>{ hasError }</p>}
    </Container>
  );
}

export default Home;