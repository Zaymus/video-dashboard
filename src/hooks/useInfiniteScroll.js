import { useEffect, useRef, useCallback } from 'react';
import useDebounce from "./useDebounce";

const useInfiniteScroll = ({ callback, delay=200, offset=0, isLoading=false, targetRef }) => {
  const callbackRef = useRef(callback);

  useEffect(() => {
    callbackRef.current = callback;
  }, [callback]);

  const scrollHandler = useCallback(() => {
    if (isLoading) return;

    const target = targetRef?.current;
    const scrollTop = target ? target.scrollTop : window.scrollY;
    const clientHeight = target ? target.clientHeight : window.innerHeight;
    const scrollHeight = target ? target.scrollHeight : document.documentElement.scrollHeight;

    if (scrollTop + clientHeight >= scrollHeight - offset) {
      callbackRef.current();
    }
  }, [isLoading, offset, targetRef]);

  const debouncedScrollHandler = useDebounce(scrollHandler, delay);

  useEffect(() => {
    let target = targetRef?.current || window;
    let observer;

    if (targetRef && !target.current) {
      observer = new MutationObserver(() => {
        if (targetRef.current) {
          target = targetRef.current;
          target.addEventListener("scroll", debouncedScrollHandler);
          observer.disconnect();
        }
      });

      observer.observe(document.body, { childList: true, subtree: true });
    } else {
      target.addEventListener("scroll", debouncedScrollHandler);
    }

    return () => {
      if (observer) observer.disconnect();
      if (target) target.removeEventListener("scroll", debouncedScrollHandler);
    };
  }, [targetRef, debouncedScrollHandler]);
}

export default useInfiniteScroll;