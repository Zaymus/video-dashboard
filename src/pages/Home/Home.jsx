import { useEffect, useState, Suspense, useCallback } from 'react';
import Loader from '../../components/common/Loader';
import VideoCard from '../../components/VideoCard';
import useAPI from '../../hooks/useAPI';
import { API_ENDPOINTS, POPULAR_VIDEOS_REQUEST } from '../../utils/constants';
import styled from 'styled-components';
import { Fullscreen } from '../../components/common';
import useInfiniteScroll from '../../hooks/useInfiniteScroll';
import useScreenSize from '../../hooks/useScreenSize';

const VideosContainer = styled.div`
  display: flex;
  width: 100%;
  flex-wrap: wrap;
`;

const Home = () => {
  const [videos, setVideos] = useState([]);
  const [nextPageToken, setNextPageToken] = useState(null);
  const [requestConfig, setRequestConfig] = useState(POPULAR_VIDEOS_REQUEST);
  const { result, isLoading, hasError } = useAPI(API_ENDPOINTS.getVideos, requestConfig);
  const screenSize = useScreenSize();

  useEffect(() => {
    if (result && result.nextPageToken !== nextPageToken) {
      setNextPageToken(result.nextPageToken);
      setVideos(prev => [...prev, ...result.items]);
    }
  }, [result])

  const loadNextPage = useCallback(() => {
    if(!isLoading && nextPageToken) {
      setRequestConfig(prevConfig => {
        if (prevConfig.params?.pageToken === nextPageToken) return prevConfig;

        return {
          ...prevConfig,
          params: {
            ...prevConfig.params,
            pageToken: nextPageToken
          }
        };
      });
    }
  }, [isLoading, nextPageToken]);

  useInfiniteScroll({
    callback: loadNextPage,
    delay: 300,
    isLoading
  });

  return (
    <Suspense fallback={<Fullscreen className="fitHeader"><Loader /></Fullscreen>}>
      <VideosContainer>
        {
          videos.length > 0 && videos.map(video => {
            return <VideoCard 
              id={video.id}
              title={video.snippet.title}
              thumbnails={video.snippet.thumbnails}
              cardsPerRow={screenSize === 'mobile' ? 1 : screenSize === 'tablet' ? 3 : 5}
            />;
          })
        }
        { isLoading == true && <Loader /> }
        { hasError && <p>{ hasError }</p>}
      </VideosContainer>
    </Suspense>
  );
}

export default Home;