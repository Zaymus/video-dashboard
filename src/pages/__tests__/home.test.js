import { jest, describe, test, expect, beforeEach } from '@jest/globals';
import '@testing-library/jest-dom';

jest.mock('../../hooks/useAPI');
jest.mock('../../utils/env');
jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useNavigate: () => jest.fn(),
}));

import Home from '../Home';
import { render, screen, act, waitFor } from '@testing-library/react';
import { setUseAPIMock, loadingMock, errorMock, defaultResult, moreVideosMock, responsiveMock } from '../../hooks/useAPI';

Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 1920 });

describe('Home Page', () => {
  beforeEach(() => {
    window.innerWidth = 1920;
    jest.resetAllMocks();
  })

  test('renders videos from useAPI result', async () => {
    setUseAPIMock(undefined);
    const { container } = render(<Home />);

    await waitFor(() => {
      defaultResult.result.items.forEach(video => {
        expect(screen.getByText(video.snippet.title)).toBeInTheDocument();
      });
    });
    expect(container).toMatchSnapshot();
  });

  test('shows loader when isLoading is true', () => {
    setUseAPIMock(loadingMock);

    const { container } = render(<Home />);
    expect(screen.getByRole('progressbar')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('shows error message when hasError is set', () => {
    setUseAPIMock(errorMock);

    const { container } = render(<Home />);
    expect(screen.getByText('API error')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

   test('renders additional videos from useAPI result when scrollbar is at bottom', async () => {
    setUseAPIMock(undefined);
    const { container } = render(<Home />);

    await waitFor(() => {
      defaultResult.result.items.forEach(video => {
        expect(screen.getByText(video.snippet.title)).toBeInTheDocument();
      });
    });

    const scrollContainer = screen.getByTestId('video-container');
    
    act(() => {
      setUseAPIMock(moreVideosMock);
      scrollContainer.dispatchEvent(new Event("scroll"));
    });

    await waitFor(() => {
      moreVideosMock.result.items.forEach(video => {
        expect(screen.getByText(video.snippet.title)).toBeInTheDocument();
      });
    });
    expect(container).toMatchSnapshot();
  });

  test('Matches snapshot in small mobile view', async () => {
    setUseAPIMock(responsiveMock);
    window.innerWidth = 375;
    const { container } = render(<Home />);

    await waitFor(() => {
      responsiveMock.result.items.forEach(video => {
        expect(screen.getByText(video.snippet.title)).toBeInTheDocument();
      });
    });
    expect(container).toMatchSnapshot();
  });

  test('Matches snapshot in large mobile view', async () => {
    setUseAPIMock(responsiveMock);
    window.innerWidth = 425;
    const { container } = render(<Home />);

    await waitFor(() => {
      responsiveMock.result.items.forEach(video => {
        expect(screen.getByText(video.snippet.title)).toBeInTheDocument();
      });
    });
    expect(container).toMatchSnapshot();
  });

  test('Matches snapshot in tablet view', async () => {
    setUseAPIMock(responsiveMock);
    window.innerWidth = 768;
    const { container } = render(<Home />);

    await waitFor(() => {
      responsiveMock.result.items.forEach(video => {
        expect(screen.getByText(video.snippet.title)).toBeInTheDocument();
      });
    });
    expect(container).toMatchSnapshot();
  });
});