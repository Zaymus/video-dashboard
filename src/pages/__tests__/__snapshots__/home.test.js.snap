// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home Page renders additional videos from useAPI result when scrollbar is at bottom 1`] = `
<div>
  <div
    class="sc-hvigdm gdIrIS"
    id="homeContainer"
  >
    <p>
      Video 1
    </p>
    <p>
      Video 2
    </p>
    <p>
      Video 3
    </p>
    <p>
      Video 4
    </p>
    <p>
      Video 5
    </p>
    <p>
      Video 6
    </p>
  </div>
</div>
`;

exports[`Home Page renders videos from useAPI result 1`] = `
<div>
  <div
    class="sc-hvigdm gdIrIS"
    id="homeContainer"
  >
    <p>
      Video 1
    </p>
    <p>
      Video 2
    </p>
    <p>
      Video 3
    </p>
  </div>
</div>
`;

exports[`Home Page shows error message when hasError is set 1`] = `
<div>
  <div
    class="sc-hvigdm gdIrIS"
    id="homeContainer"
  >
    <p>
      API error
    </p>
  </div>
</div>
`;

exports[`Home Page shows loader when isLoading is true 1`] = `
<div>
  <div
    class="sc-hvigdm gdIrIS"
    id="homeContainer"
  >
    <div
      class="sc-bRKDuR fMrgJQ"
      role="progressbar"
    >
      <svg
        class="spinner"
        viewBox="0 0 50 50"
      >
        <circle
          class="path"
          cx="25"
          cy="25"
          fill="none"
          r="20"
          stroke-width="5"
        />
      </svg>
    </div>
  </div>
</div>
`;
